balancer/weightedroundrobin/balancer.go

```diff
@@ -0,0 +1,493 @@
+	if w.logger.V(2) {
+		w.logger.Infof("wrr: new weight for subchannel %v: %v", w.SubConn, w.weightVal)
+	}
+
+	w.lastUpdated = time.Now()
+	if w.nonEmptySince == (time.Time{}) {
+		w.nonEmptySince = w.lastUpdated
+	}
+}
+
+func (w *weightedSubConn) updateConfig(cfg *lbConfig) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	oldCfg := w.cfg
+	if oldCfg == nil {
+		oldCfg = &lbConfig{EnableOOBLoadReport: false}
+	}
+	w.cfg = cfg
+	newPeriod := cfg.OOBReportingPeriod
+	if cfg.EnableOOBLoadReport == oldCfg.EnableOOBLoadReport &&
+		newPeriod == oldCfg.OOBReportingPeriod {
+		// Load reporting wasn't enabled before or after, or load reporting was
+		// enabled before and after, and had the same period.  (Note that with
+		// load reporting disabled, OOBReportingPeriod is always 0.)
+		return
+	}
+	// (Optionally stop and) start the listener.
+	if w.stopORCAListener != nil {
+		w.stopORCAListener()
+	}
+	if !cfg.EnableOOBLoadReport {
+		w.stopORCAListener = nil
+		return
+	}
+	w.logger.Infof("Registering listener for %v; %v", w.SubConn, newPeriod)
+	opts := orca.OOBListenerOptions{ReportInterval: newPeriod}
+	w.stopORCAListener = orca.RegisterOOBListener(w.SubConn, w, opts)
+}
+
+func (w *weightedSubConn) updateConnectivityState(cs connectivity.State) connectivity.State {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	switch cs {
+	case connectivity.Idle:
+		// Always reconnect when idle.
+		w.SubConn.Connect()
+	case connectivity.Ready:
+		// If we transition back to READY state, restart the blackout period.
+		// Note that we cannot guarantee that we will never receive lingering
+		// callbacks for backend metric reports from the previous connection
+		// after the new connection has been established, but they should be
+		// masked by new backend metric reports from the new connection by the
+		// time the blackout period ends.
+		w.nonEmptySince = time.Time{}
+	case connectivity.Shutdown:
+		if w.stopORCAListener != nil {
+			w.stopORCAListener()
+		}
+	}
+
+	old := w.connectivityState
+
+	if old == connectivity.TransientFailure &&
+		(cs == connectivity.Connecting || cs == connectivity.Idle) {
+		// Once a subconn enters TRANSIENT_FAILURE, ignore subsequent IDLE or
+		// CONNECTING transitions to prevent the aggregated state from being
+		// always CONNECTING when many backends exist but are all down.
+		return old
+	}
+
+	w.connectivityState = cs
+
+	return old
+}
+
+func (w *weightedSubConn) weight(now time.Time, weightExpirationPeriod, blackoutPeriod time.Duration) float64 {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	// If the most recent update was longer ago than the expiration period,
+	// reset nonEmptySince so that we apply the blackout period again if we
+	// start getting data again in the future, and return 0.
+	if now.Sub(w.lastUpdated) > weightExpirationPeriod {
+		w.nonEmptySince = time.Time{}
+		return 0
+	}
+	// If we don't have at least blackoutPeriod worth of data, return 0.
```

Comment:
This is just stating the code. As per top level function comment, I would prefer reasoning as to root cause for why we return 0 and what returning 0 implies and why.